### DISCUSSION TITLE
How do I make MIDI class global across files?

### DISCUSSION THREAD

**Discussion by fapplin** [earlephilhower/arduino-pico#737] [General]
I use this statement at the top of my main file (with setup() Loop()):

MIDI_CREATE_INSTANCE(Adafruit_USBD_MIDI, usb_midi, MIDI);

How do I make the "MIDI" global, so I can use it in other files?

Thanks.

**Comment by Andy2No** [comment]
> How do I make the "MIDI" global, so I can use it in other files?

I think it should just work, if the other tabs are also .ino files.  Apparently, the compiler concatenates all .ino files, starting with the one that matches the folder/project name, and compiles them as though they were one big long file.  So, anything declared in the main one should be visible to the others, though not the other way around - you'd need to declare function prototypes in the main .ino, to be able to use functions declared in the others, for example. [EDIT: It seems you may only have to do that for functions with default values, i.e. optional parameters.  The IDE should generate the rest automatically.  See https://forum.arduino.cc/t/default-argument/885323/6]

There are other approaches, like using #include "another_tab.h", to just insert another chunk of the program into the main tab -. h files don't have to be header files, as such, they're just inserted where the **#include** statement is.  You can also make the other tabs as .cpp files, and use **extern** to refer to things in them.

This thread covers some of that, but maybe not all the details:

https://forum.arduino.cc/t/how-to-properly-include-functions-written-on-other-sketch-tabs/565672

**Reply by fapplin** [reply]
I feel like this MIDI object is part of the core (even if it is from Adafruit). So, it should be an appropriate place to ask the question.

I have a globals.cpp and globals.h file. This is where I define all of my global variables that are used among the project files. I have a midicalls.cpp and  a midicalls.h, also. I wanted to separate out the noteon, noteoff, pitchbend, etc. from the main .ino file. So, I need to know how to set up the MIDI object used in:

MIDI_CREATE_INSTANCE(Adafruit_USBD_MIDI, usb_midi, MIDI);

so, I can use it wherever I want. If this should be directed towards the Adafruit crowd - I will do so.

Thanks.

**Reply by ArcoK** [reply]
A little late, but I just recently ran into this same problem. 
This is what I did, tested it and it works , hope some people find this useful.
`
/*	=======================	
	Main .ino file:
	=======================	*/

#include <Arduino.h>
#include <Adafruit_TinyUSB.h>
#include "<MIDI.h>"

// USB MIDI object
Adafruit_USBD_MIDI usb_midi;

// Create a new instance of the Arduino MIDI Library,
// and attach usb_midi as the transport.
MIDI_CREATE_INSTANCE(Adafruit_USBD_MIDI, usb_midi, MIDI);
// create a pointer to this thing
midi::MidiInterface<midi::SerialMIDI<Adafruit_USBD_MIDI> >* pMidi = &MIDI;


#include "midiProcessing.h"	// class for midi stuff

MidiProcessor processor(pMidi);	// create object and pass the pointer

void setup() 
{
	// setup stuff
}

void loop()
{
	processor.doMidiStuff();
}
'

`
/*	=======================	
	file: midiProcessing.h
	=======================	*/
#include <Adafruit_TinyUSB.h>
#include <MIDI.h>

class MidiProcessor
{
  public:
	midi::MidiInterface<midi::SerialMIDI<Adafruit_USBD_MIDI> >* myMidi;
	/** current midi message that is played, previous that must be 'noted off' */
  	midi::Message<0> msgCurrent, msgPrevious;

	/** pass pointer to MIDI object in constructor */
	VqProcessor(midi::MidiInterface<midi::SerialMIDI<Adafruit_USBD_MIDI> >* p);
	~VqProcessor();

	void doMidiStuff()
	{
		/** When appropriate time has elapsed: */
		myMidi->sendNoteOff(msgPrevious.data1 , 0, msgPrevious.channel);
	    msgCurrent.data1 = random(36, 102);;
	    myMidi->sendNoteOn(msgCurrent.data1 , 100, msgCurrent.channel);
	    msgPrevious = msgCurrent;
	}

};

`

**Comment by JAndrassy** [comment]
put
```
extern midi::MidiInterface<Adafruit_USBD_MIDI> midiA;
```
into globals.h

**Reply by fapplin** [reply]
OK. I'm making some assumptions (which is never good). I'm assuming

extern midi::MidiInterface<Adafruit_USBD_MIDI> midiA;

the midiA is the object I will use. So, I made the following changes:

MIDI_CREATE_INSTANCE(Adafruit_USBD_MIDI, usb_midi, midiA);

and made the appropriate changes like:

midiA.sendNoteOn(pitch, velocity, channel);

etc.

I get the following error:

conflicting declaration 'midi::MidiInterface<midi::SerialMIDI<Adafruit_USBD_MIDI> > midiA'

What am I doing wrong?

**Reply by Andy2No** [reply]
>  get the following error:
>
> conflicting declaration 'midi::MidiInterface<midi::SerialMIDI<Adafruit_USBD_MIDI> > midiA'

There should be a line after that that says which file the error occurred in, and on what line.

I think we'd need to see more of the code, or a stripped down version of it that still produces the same error.

**Reply by fapplin** [reply]
OK. I made a stripped down version. So, here are my files.

test-01.ino
```
#include <Arduino.h>
#include <Adafruit_TinyUSB.h>
#include <MIDI.h>
#include "globals.h"
#include "midicalls.h"


// USB MIDI object
//Adafruit_USBD_MIDI usb_midi;

// Create a new instance of the Arduino MIDI Library,
// and attach usb_midi as the transport.
MIDI_CREATE_INSTANCE(Adafruit_USBD_MIDI, usb_midi, myMidy);

void MidiIn() {
  //Serial.println(Serial2.available());
  if (Serial2.available() > 0) {
    // read the incoming byte:
    byte incomingByte = Serial2.read();
    //Serial.println(incomingByte);
    switch (state) { // this reads all incoming bytes, realtime stuff is ignored
      case 0:
        if (incomingByte > 0x7F && incomingByte < 0xF0 ) {
          statusByte = incomingByte;
          state = 1;
        }
        if (incomingByte == 0xF0)
          state = 3;
      case 1:
        // get first databyte
        if (incomingByte < 0x80) {
          dataByte1 = incomingByte;
          //check for two byte messages...
          if ((statusByte & 0xF0) == 0xC0 | (statusByte & 0xF0) == 0xD0) {
            midiParser();
            state = 0; //reset state machine
          } else
            state = 2;
        } else {
          state = 0;  // reset state machine as this should be a note number
        }
        break;

      case 2:
        // get second databyte
        if (incomingByte < 128) {
          dataByte2 = incomingByte;
        }
        midiParser();
        state = 0;  // reset state machine to start
        break;
      //stay here for sysex, wait for it to end...
      case 3:
        if (incomingByte == 0xF7) state = 0;
        if (incomingByte > 0x7F && incomingByte < 0xF0) {
          statusByte = incomingByte;
          state = 1;
        }
        break;
    }
  }
}
void midiParser() {
  //  Serial.println("=============================");
  //  Serial.print("statusByte: ");
  //  Serial.println(statusByte);
  //  Serial.print("dataByte1: ");
  //  Serial.println(dataByte1);
  //  Serial.print("dataByte2: ");
  //  Serial.println(dataByte2);
  //  Serial.println("=============================");
  switch (statusByte) {
    case 0x90:
      if (dataByte2 > 0) {
        //        if (dataByte2 == prev_note)
        //          break;

        doNoteOn(1, dataByte1, dataByte2);
      } else {
        doNoteOff(1, dataByte1, 0);
      } //end if

      break;
    case 0xE0: //pitch bend
      //unsigned short pitch_value;
      //pitch_value= CombineBytes(dataByte1, dataByte2);
      //doPitchBend(1, pitch_value);
      break;
    case 0xC0: //program change
      //do something when a program change arrives, prg nr is dataByte1
      break;
    case 0xB0: //cc messages
      switch (dataByte1) {
        case SAVE_CC:
          //do something here
          break;
        case CHANNELMODE_CC:
        //do something else
        case STICK_CC:
          //again something on this cc
          break;
        case FULLLEGATO_CC:
          //you get the idea
          break;
        case OCTAVE_CC:
          //octaves!!!
          break;
      } //end switch
  } //end switch
}

void setup() {
#if defined(ARDUINO_ARCH_MBED) && defined(ARDUINO_ARCH_RP2040)
  // Manual begin() is required on core without built-in support for TinyUSB such as mbed rp2040
  TinyUSB_Device_Init(0);
#endif
 

  myMidy.begin(MIDI_CHANNEL_OMNI);
  myMidy.turnThruOff();

  Serial2.setRX(5);
  Serial2.setTX(4);
  Serial2.begin(31250);

  Serial.begin(9600);

  while ( !TinyUSBDevice.mounted() ) delay(1);

}

void loop() {

  //Serial.println("looping");
  MidiIn();

}
```
globals.cpp
```
#include "globals.h"
#include <Adafruit_TinyUSB.h>
#include <MIDI.h>

#define SAVE_CC 120
#define CHANNELMODE_CC 119
#define FULLLEGATO_CC 118
#define OCTAVE_CC 117
#define STICK_CC 116

byte state = 0;
byte statusByte;
byte dataByte1;
byte dataByte2;

Adafruit_USBD_MIDI usb_midi;
```

globals.h
```
#ifndef GLOBALS_H

#define GLOBALS_H

#include <Arduino.h>
#include <Adafruit_TinyUSB.h>
#include <MIDI.h>

#define SAVE_CC 120
#define CHANNELMODE_CC 119
#define FULLLEGATO_CC 118
#define OCTAVE_CC 117
#define STICK_CC 116

extern byte state;
extern byte statusByte;
extern byte dataByte1;
extern byte dataByte2;

extern Adafruit_USBD_MIDI usb_midi;
extern midi::MidiInterface<Adafruit_USBD_MIDI> myMIDI;


#endif
```
midicalls.cpp
```
#include "globals.h"
#include "midicalls.h"

void doNoteOn(byte channel, byte pitch, byte velocity)
{
    myMIDI.sendNoteOn(pitch, velocity, channel);
}

void doNoteOff(byte channel, byte pitch, byte velocity)
{
    myMIDI.sendNoteOn(pitch, 0, channel);
}
```

midicalls.h
```
#ifndef MIDICALLS_H

#define MIDICALLS_H


extern "C" void doNoteOn(byte channel, byte pitch, byte velocity);
extern "C" void doNoteOff(byte channel, byte pitch, byte velocity);

#endif
```

Here's the errors.
```
Arduino: 1.8.15 (Linux), TD: 1.54, Board: "Raspberry Pi Pico W, 2MB (Sketch: 1920KB, FS: 128KB), 133 MHz, Small (-Os) (standard), Disabled, Disabled, Disabled, Disabled, None, Adafruit TinyUSB, IPv4 Only"











In file included from /home/frank/Arduino/libraries/MIDI_Library/src/MIDI.h:307,
                 from /home/frank/Arduino/test-01/globals.h:7,
                 from /home/frank/Arduino/test-01/midicalls.cpp:1:
/home/frank/Arduino/libraries/MIDI_Library/src/MIDI.hpp: In instantiation of 'void midi::MidiInterface<Transport, _Settings, _Platform>::send(midi::MidiType, midi::DataByte, midi::DataByte, midi::Channel) [with Transport = Adafruit_USBD_MIDI; _Settings = midi::DefaultSettings; _Platform = midi::DefaultPlatform; midi::DataByte = unsigned char; midi::Channel = unsigned char]':
/home/frank/Arduino/libraries/MIDI_Library/src/MIDI.hpp:230:9:   required from 'void midi::MidiInterface<Transport, _Settings, _Platform>::sendNoteOn(midi::DataByte, midi::DataByte, midi::Channel) [with Transport = Adafruit_USBD_MIDI; _Settings = midi::DefaultSettings; _Platform = midi::DefaultPlatform; midi::DataByte = unsigned char; midi::Channel = unsigned char]'
/home/frank/Arduino/test-01/midicalls.cpp:6:47:   required from here
/home/frank/Arduino/libraries/MIDI_Library/src/MIDI.hpp:180:24: error: 'class Adafruit_USBD_MIDI' has no member named 'beginTransmission'
  180 |         if (mTransport.beginTransmission(inType))
      |             ~~~~~~~~~~~^~~~~~~~~~~~~~~~~
/home/frank/Arduino/libraries/MIDI_Library/src/MIDI.hpp:204:24: error: 'class Adafruit_USBD_MIDI' has no member named 'endTransmission'
  204 |             mTransport.endTransmission();
      |             ~~~~~~~~~~~^~~~~~~~~~~~~~~
Multiple libraries were found for "Adafruit_TinyUSB.h"
 Used: /home/frank/Arduino/libraries/Adafruit_TinyUSB_Library
 Not used: /home/frank/.arduino15/packages/rp2040/hardware/rp2040/2.3.3/libraries/Adafruit_TinyUSB_Arduino
exit status 1
Error compiling for board Raspberry Pi Pico W.


This report would have more information with
"Show verbose output during compilation"
option enabled in File -> Preferences.
```

**Comment by Andy2No** [comment]
Not a complete answer, but I think you need to swap the order of these two in globals.cpp :
```
#include "globals.h"
#include <Adafruit_TinyUSB.h>
```

Also, globals.h has the extern declarations (i.e. defining the references as being external), so you don't want that part included in the .ino, where those two things are defined.  In C++ you only do that for const definitions, apparently.

You could either split globals.h into two parts, or use a #define to say this is the file where they're actually defined, and a #ifndef ... #endif in globals.h, to stop it doing that part when it's included in the .ino.

Also, unless I'm missing something, you need to uncomment this line, so you're actually defining the usb_midi object in the .ino file:

```
//Adafruit_USBD_MIDI usb_midi;
```

**Comment by Andy2No** [comment]
Here's a couple of references on how extern is used in C++, which should also apply here:

https://stackoverflow.com/questions/10422034/when-to-use-extern-in-c

https://docs.microsoft.com/en-us/cpp/cpp/extern-cpp?view=msvc-170

Basically, in C++ you use extern in any module where you want to refer to a variable that's in another module (that's called a declaration) but you can't use it in the module where the variable is defined... unless it's a const, for some reason.  The rule is different in C, which lets you have extern (as a declaration) followed by a definition of the variable - so you'd get to use the same include file in each module.

**Reply by fapplin** [reply]
Thanks. This was initially in C. I tried your suggestions and still no luck. More trouble than it's worth for the few routines I'm trying to separate out. Thanks for the help.

**Reply by TijuanaKez** [reply]
Did you ever end up getting to the bottom of this?
I'm having the exact same problem; have lot's of midi functions so they're in their own h/cpp and I have globals.h. Getting the same errors about mTransport.beginTransmission etc.

**Reply by fapplin** [reply]
I ended up using a different MIDI library and I made a Class in a .cpp file to do a lot of my MIDI work. So, the callback stuff is done in the Main .ino file, but I hand off work to the Class.

https://github.com/FortySevenEffects/arduino_midi_library

**Comment by Andy2No** [comment]
No problem.  I had success converting a multi module C program to Arduino a while ago, by leaving most of it as C and using **extern "C" ...** in the ,ino main module to refer to global variables and functions in the C modules.

This is a little different though because you're using the C++ library.

I'd be tempted to go with one big .ino too, and make use of the folding editor functions of the IDE, or break it up into multiple .ino modules using the rules explained in one of the first links I gave, in an earlier post - the compiler just pastes them all together, starting with the one matching the folder name, then apparently in alphabetical order.   That might be more convenient than the one big .ino approach, but it amounts to the same thing, really.   There would be no difference in compile time, for one thing.
